src/hyp_map.rs

```diff
@@ -310,11 +314,12 @@ impl HypPageTable {
             UmodeSlotId::A => 0,
             UmodeSlotId::B => 1,
         };
-        // Unwrap okay: the result is dependent on constant that must be page aligned.
-        PageAddr::new(RawAddr::supervisor_virt(
-            UMODE_MAPPINGS_START + slot_num * UMODE_MAPPING_SLOT_SIZE,
-        ))
-        .unwrap()
+        // UMODE_MAPPING_START and UMODE_MAPPING_SLOT_SIZE are required by static assertion to be
+        // page aligned, `round_down` will be a nop.
+        PageAddr::with_round_down(
+            RawAddr::supervisor_virt(UMODE_MAPPINGS_START + slot_num * UMODE_MAPPING_SLOT_SIZE),
+            PageSize::Size4k,
+        )
```

Comment:
Yes, This was my attempt at reproducing changes as started here: https://github.com/rivosinc/salus/pull/212/files

And wanted to go with something that follows along the same lines so they can all be changed later, when we find a better interface.

Reply:
I hate unwraps because they almost always signal we aren't using the defined types correctly. And in this case a const PageAddr is the 'correct' fix. But that's a pretty tough exercise as you can't declare the `AS` type parameter as const. I'm OK landing this either way and we can fight the const constructor battle separately.

Reply:
Yes please. Can we please merge this?

Reply:
Yes, I'm good with merging this one. Getting const page addrs was not fun, but I got something to work: https://github.com/dgreid/salus/commits/topic/const_page_addr

I can rebase after this lands.

Reply:
And it turns out Dylan is way ahead of me again :) Let me comment over there.

Reply:
Came up with this macro. It does use unwrap (alternative is to use an unsafe function to create a raw page addr), but at least it's contained in a const_assert.

We could use this in the future. 

```
index d4431b4..f790591 100644
--- a/riscv-pages/src/page.rs
+++ b/riscv-pages/src/page.rs
@@ -169,6 +169,22 @@ pub struct PageAddr<AS: AddressSpace> {
     addr: RawAddr<AS>,
 }
 
+/// Create a PageAddr from an aligned constant.
+///
+/// Usage example:
+///
+///     use riscv_pages::*;
+///
+///     let x = const_page_addr!(0x1000, PageSize::Size4k, SupervisorVirt::default());
+#[macro_export]
+macro_rules! const_page_addr {
+    ($addr: expr, $page_size: expr, $address_space: expr) => ({
+        static_assertions::const_assert!($page_size.is_aligned($addr));
+        // Unwrap okay: static assertion right above ensures aslignment.
+        PageAddr::with_alignment(RawAddr::new($addr, $address_space), PageSize::Size4k).unwrap();
+    })
+}
+
 /// A page-aligned address in the `SupervisorPhys` address space.
 pub type SupervisorPageAddr = PageAddr<SupervisorPhys>;
 /// A page-aligned address in a `GuestPhys` address space.
@@ -781,6 +797,13 @@ mod tests {
         .is_some());
     }
 
+    #[test]
+    fn const_test() {
+        let _x = const_page_addr!(0x1000, PageSize::Size4k, SupervisorVirt::default());
+        let _x = const_page_addr!(0x200000, PageSize::Size2M, SupervisorVirt::default());
+        let _x = const_page_addr!(0x40000000, PageSize::Size1G, SupervisorVirt::default());
+    }
+
     #[test]
     fn round_phys() {
         assert!(

```

Reply:
I am actually okay with adding this after it lands, or just importing your patch in the PR. I don't have strong opinion. I tried to do something useful adding a change that removed the previous unwraps with something similar that what happened before (plus adding const_assert).

The unfortunate timing is that the patch that introduced the unwrap above went in at the same time as your patch removing other unwraps, so I tried to fix this up.